### DISCUSSION TITLE
missing nvidia_drv.so

### DISCUSSION THREAD

**Discussion by phirestalker** [games-on-whales/gow#57] [Q&A]
In the documentation it mentions mapping the directory containing nvidia_drv.so to the Xorg container if you have the Nvidia drivers. I do have the Nvidia drivers, but I do not seem to have the nvidia_drv.so file. I am on a server install of Ubuntu 20.04. I did install the server Nvidia drivers. I do not want a graphical environment on my host machine, so I removed the regular drivers as they required X11 packages.
nvidia-smi works fine as does a test container with the Nvidia runtime.
So, how do I get it working without this file, or how do I get this file without any X11 stuff on the host machine?

**Comment by ABeltramo** [comment]
Hello @phirestalker thanks for trying out GOW! :+1: 

I'm glad you read thru the docs before trying it out, unfortunately they are a little bit outdated on the nvidia front. Our Xorg Docker container on startup will try to automatically download the required packages, that matches the host driver versions, using APT (here's the relevant [script](https://github.com/games-on-whales/gow/blob/e6e5598bfcd8220078f3dcca7d1c13bb665492cc/images/xorg/scripts/ensure-nvidia-xorg-driver.sh#L36) in case you are interested).

The solution in the docs of passing the host drivers using a volume mount is now considered a failover in case APT is not able to match the host driver version. Have you tried running the Xorg docker container?  
Make sure to setup also `DOCKER_RUNTIME=nvidia` on the [.env file](https://github.com/games-on-whales/gow/blob/e6e5598bfcd8220078f3dcca7d1c13bb665492cc/.env#L31)

**Comment by phirestalker** [comment]
Yes, I did try running it. The web interface was not available, so I started checking the docker logs for each container. I found an error about not finding a display. I was going to try and troubleshoot it, but it would not shut down. I tried using down multiple times and finally had to use kill.
I would have continued troubleshooting, but starting the containers again gave errors and ended up freezing my entire server.
I was just now thinking that since the script checks first for the mounted drivers that it was the preferred method and might work better.

I should have written down which containers did not want to quit. I figured I would just wait for another release first.

I did set nvidia in the env file.
I am a little confused as to what the Xorg container is doing since it doesn't use the nvidia runtime. Is there somewhere I can read more about what is happening behind the scenes?

**Reply by ABeltramo** [reply]
I'm sorry to hear that, freezing the server it's a severe error, would you like to share more about your HW+OS? Are you running it inside a VM or on bare metal? 

> I found an error about not finding a display

Could it be that you don't have a plugged monitor in your server (a dummy HDMI dongle would be fine too). In order to start Xorg a video output is required AFAIK.

**Reply by phirestalker** [reply]
Sorry I did not reply correctly:

My "server" is an old 2010 Mac Pro tower running Ubuntu 20.04. That host is where I start the docker containers from. It has 32 GB of RAM with only about 9 used. My graphics card is a GeForce GTX 680. Let me know any more details that you need.

I was able to get it to start and find the display (I must have had the cable disconnected before).
I am still having trouble shutting the containers down. So far I have only tried using the down command with docker-compose, but it is hanging and timing out on the inspect_container command. I am unsure which container it is.

EDIT:
it seems both the retroarch and xorg containers refuse to quit even with a docker kill command.

**Comment by Hokuten** [comment]
I found Nvidia_drv.so by extracting it from the nvidia.run binary downloaded from Nvidia. I.E. nvidia.run -x --target /DIR and it will put the library in that /DIR forlder.
